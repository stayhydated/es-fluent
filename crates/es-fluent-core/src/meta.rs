//! This module provides types for representing the kind of a type.

use serde::Serialize;

#[derive(Clone, Debug)]
pub struct EnumKind;

#[derive(Clone, Debug)]
pub struct StructKind;

#[derive(Clone, Debug)]
pub struct StructKvKind;

#[derive(Clone, Debug, strum::Display, strum::IntoStaticStr, Eq, Hash, PartialEq, Serialize)]
#[strum(serialize_all = "snake_case")]
pub enum TypeKind {
    Enum,
    Struct,
}
