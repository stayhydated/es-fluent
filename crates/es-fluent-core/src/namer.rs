use derive_more::{Debug, Deref, Display};
use heck::ToSnakeCase as _;
use quote::format_ident;

#[derive(Clone, Debug, Deref, Display, Eq, Hash, Ord, PartialEq, PartialOrd)]
pub struct FluentKey(pub String);

impl FluentKey {
    pub const DELIMITER: &str = "-";
    pub fn new(ftl_name: &syn::Ident, sub_name: &str) -> Self {
        let normalized_name = ftl_name.to_string().to_snake_case();
        if sub_name.is_empty() {
            FluentKey(normalized_name)
        } else {
            FluentKey(format!(
                "{}{}{}",
                normalized_name,
                Self::DELIMITER,
                sub_name
            ))
        }
    }
}

impl quote::ToTokens for FluentKey {
    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
        let key_string = &self.0;
        tokens.extend(quote::quote! { #key_string });
    }
}

#[derive(Clone, Debug, Deref, Display, Eq, Hash, PartialEq)]
pub struct FluentDoc(pub String);

impl From<&FluentKey> for FluentDoc {
    fn from(ftl_key: &FluentKey) -> Self {
        FluentDoc(format!("Key = `{}`", *ftl_key))
    }
}

impl quote::ToTokens for FluentDoc {
    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
        let doc_string = &self.0;
        tokens.extend(quote::quote! { #doc_string });
    }
}

pub struct UnnamedItem(usize);

impl std::fmt::Display for UnnamedItem {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}{}", Self::UNNAMED_PREFIX, self.0)
    }
}

impl UnnamedItem {
    const UNNAMED_PREFIX: &str = "f";

    pub fn to_ident(&self) -> syn::Ident {
        format_ident!("{}", self.to_string())
    }
}

impl From<usize> for UnnamedItem {
    fn from(index: usize) -> Self {
        Self(index)
    }
}
