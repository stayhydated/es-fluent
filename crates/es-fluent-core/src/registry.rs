use crate::meta::TypeKind;
use crate::namer::FluentKey;
use bon::Builder;

#[derive(Builder, Clone, Debug, Eq, Hash, PartialEq)]
pub struct FtlVariant {
    pub name: String,
    pub ftl_key: FluentKey,
    pub arguments: Option<Vec<String>>,
}

#[derive(Builder, Clone, Debug, Eq, Hash, PartialEq)]
pub struct FtlTypeInfo {
    pub type_kind: TypeKind,
    pub type_name: String,
    pub variants: Vec<FtlVariant>,
}
