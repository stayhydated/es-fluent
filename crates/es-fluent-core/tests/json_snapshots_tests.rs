use es_fluent_core::meta::TypeKind;
use es_fluent_core::namer::FluentKey;
use es_fluent_core::registry::{FtlTypeInfo, FtlVariant};
use proc_macro2::Span;
use syn::Ident;

#[test]
fn json_snapshot_ftl_variant() {
    let ident = Ident::new("ErrorType", Span::call_site());

    let variant = FtlVariant::builder()
        .name("Network".to_string())
        .ftl_key(FluentKey::new(&ident, "Network"))
        .maybe_args(Some(vec!["code".into(), "message".into()]))
        .build();

    insta::assert_json_snapshot!("json_snapshot_ftl_variant", &variant);
}

#[test]
fn json_snapshot_ftl_type_info() {
    let ident = Ident::new("ErrorType", Span::call_site());

    let variant1 = FtlVariant::builder()
        .name("Network".to_string())
        .ftl_key(FluentKey::new(&ident, "Network"))
        .maybe_args(Some(vec!["code".into(), "message".into()]))
        .build();

    let variant2 = FtlVariant::builder()
        .name("Io".to_string())
        .ftl_key(FluentKey::new(&ident, "Io"))
        .maybe_args(None)
        .build();

    let type_info = FtlTypeInfo::builder()
        .type_kind(TypeKind::Struct)
        .type_name("ErrorType".to_string())
        .variants(vec![variant1, variant2])
        .build();

    insta::assert_json_snapshot!("json_snapshot_ftl_type_info", &type_info);
}
