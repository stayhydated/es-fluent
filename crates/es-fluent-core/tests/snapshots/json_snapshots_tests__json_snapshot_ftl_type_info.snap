---
source: crates/es-fluent-core/tests/json_snapshots_tests.rs
expression: "&type_info"
---
{
  "type_kind": "Struct",
  "type_name": "ErrorType",
  "variants": [
    {
      "name": "Network",
      "ftl_key": "error_type-Network",
      "args": [
        "code",
        "message"
      ]
    },
    {
      "name": "Io",
      "ftl_key": "error_type-Io",
      "args": []
    }
  ]
}
